### PR TITLE
Prevent pause a non running task and prevent resume a non paused task

### PR DESCRIPTION
**Pull request by zegnus**
## Task description ##

I have noticed that a queued download can be paused and that causes the following bug:
- Proceed downloading two batches, the first one is running, the second one is queued
- Click to pause a queued batch, queued status stays in place
- Click to pause the running batch
- The queued batch does not start as it has been paused, but as is queued it should start

## Solution implemented ##

We should only be able to pause running batches, and resume paused ones

Before | After
--- | ---
old | new

### CODE REVIEW

**Comment by hal9002** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://ci.novoda.com/job/download-manager-prb/185/
Test PASSed.

File: library/src/main/java/com/novoda/downloadmanager/lib/DownloadManager.java [hunk 1]
```diff
@@ -445,25 +445,37 @@ private long insert(Request request) {
...
+        RealSystemFacade systemFacade = new RealSystemFacade(GlobalState.getContext(), new Clock());
+        BatchRepository batchRepository = new BatchRepository(contentResolver, downloadDeleter, downloadsUriProvider, systemFacade);
 
-    public void resumeBatch(long id) {
-        ContentValues values = getResumeContentValues();
-        String where = COLUMN_BATCH_ID + "= ? AND " + DownloadContract.Downloads.COLUMN_STATUS + " != ?";
-        String[] selectionArgs = {String.valueOf(id), String.valueOf(DownloadStatus.SUCCESS)};
-        contentResolver.update(downloadsUriProvider.getAllDownloadsUri(), values, where, selectionArgs);
+        int batchStatus = batchRepository.getBatchStatus(batchId);
+        if (batchStatus == DownloadStatus.RUNNING) {
```
Comment (ouchadam) [inline library/src/main/java/com/novoda/downloadmanager/lib/DownloadManager.java]: `DownloadStatus.isRunning`?

Comment (zegnus) [inline library/src/main/java/com/novoda/downloadmanager/lib/DownloadManager.java]: Totally https://github.com/novoda/download-manager/commit/10cb8666efba80975bc5f1f71ec2c9496061b441

**Comment by hal9002** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://ci.novoda.com/job/download-manager-prb/192/
Test PASSed.

**Comment by ronocod** [comment]
Even more queries to hit the DB with :cry: but pause and resume won't be called very often so I guess it's ok for now :+1:

File: library/src/main/java/com/novoda/downloadmanager/lib/DownloadManager.java [hunk 2]
```diff
@@ -445,25 +445,37 @@ private long insert(Request request) {
...
 
-    public void pauseBatch(long id) {
-        ContentValues values = new ContentValues();
-        String where = COLUMN_BATCH_ID + "= ? AND " + DownloadContract.Downloads.COLUMN_STATUS + " != ?";
-        String[] selectionArgs = {String.valueOf(id), String.valueOf(DownloadStatus.SUCCESS)};
-        values.put(DownloadContract.Downloads.COLUMN_CONTROL, DownloadsControl.CONTROL_PAUSED);
-        contentResolver.update(downloadsUriProvider.getAllDownloadsUri(), values, where, selectionArgs);
-    }
+    public void pauseBatch(long batchId) {
+        DownloadDeleter downloadDeleter = new DownloadDeleter(contentResolver);
```
Comment (xrigau) [inline library/src/main/java/com/novoda/downloadmanager/lib/DownloadManager.java]: so awkward that we have to create a `DownloadDeleter` to pause a batch

Comment (zegnus) [inline library/src/main/java/com/novoda/downloadmanager/lib/DownloadManager.java]: Dependency of the BatchRepository :(

File: library/src/main/java/com/novoda/downloadmanager/lib/DownloadManager.java [hunk 3]
```diff
@@ -445,25 +445,37 @@ private long insert(Request request) {
...
-        String[] selectionArgs = {String.valueOf(id), String.valueOf(DownloadStatus.SUCCESS)};
-        contentResolver.update(downloadsUriProvider.getAllDownloadsUri(), values, where, selectionArgs);
+        int batchStatus = batchRepository.getBatchStatus(batchId);
+        if (DownloadStatus.isRunning(batchStatus)) {
+            ContentValues values = new ContentValues();
+            String where = COLUMN_BATCH_ID + "= ? AND " + DownloadContract.Downloads.COLUMN_STATUS + " != ?";
+            String[] selectionArgs = {String.valueOf(batchId), String.valueOf(DownloadStatus.SUCCESS)};
+            values.put(DownloadContract.Downloads.COLUMN_CONTROL, DownloadsControl.CONTROL_PAUSED);
+            contentResolver.update(downloadsUriProvider.getAllDownloadsUri(), values, where, selectionArgs);
+        }
```
Comment (xrigau) [inline library/src/main/java/com/novoda/downloadmanager/lib/DownloadManager.java]: do we just ignore ot if the batch is not running?

File: library/src/main/java/com/novoda/downloadmanager/lib/DownloadManager.java (hunk 2 above)
Comment (amlcurran) [inline library/src/main/java/com/novoda/downloadmanager/lib/DownloadManager.java]: Bit of a smell, I've noticed it before

File: library/src/main/java/com/novoda/downloadmanager/lib/DownloadManager.java (hunk 3 above)
Comment (zegnus) [inline library/src/main/java/com/novoda/downloadmanager/lib/DownloadManager.java]: Exactly, you cannot pause something that is not running :)

On Fri, 24 Jul 2015 16:59 Xavi Rigau <notifications@github.com> wrote:

> In
> library/src/main/java/com/novoda/downloadmanager/lib/DownloadManager.java
> <https://github.com/novoda/download-manager/pull/135#discussion_r35430263>
> :
>
> >
> > -    public void resumeBatch(long id) {
> > -        ContentValues values = getResumeContentValues();
> > -        String where = COLUMN_BATCH_ID + "= ? AND " + DownloadContract.Downloads.COLUMN_STATUS + " != ?";
> > -        String[] selectionArgs = {String.valueOf(id), String.valueOf(DownloadStatus.SUCCESS)};
> > -        contentResolver.update(downloadsUriProvider.getAllDownloadsUri(), values, where, selectionArgs);
> > +        int batchStatus = batchRepository.getBatchStatus(batchId);
> > +        if (DownloadStatus.isRunning(batchStatus)) {
> > +            ContentValues values = new ContentValues();
> > +            String where = COLUMN_BATCH_ID + "= ? AND " + DownloadContract.Downloads.COLUMN_STATUS + " != ?";
> > +            String[] selectionArgs = {String.valueOf(batchId), String.valueOf(DownloadStatus.SUCCESS)};
> > +            values.put(DownloadContract.Downloads.COLUMN_CONTROL, DownloadsControl.CONTROL_PAUSED);
> > +            contentResolver.update(downloadsUriProvider.getAllDownloadsUri(), values, where, selectionArgs);
> > +        }
>
> do we just ignore ot if the batch is not running?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/novoda/download-manager/pull/135/files#r35430263>.
>

-- 
Disclaimer: The information in this e-mail and any attachments is the 
property of Novoda Ltd and is confidential and may be legally privileged. 
This e-mail is intended solely for the person or organisation to which it 
is addressed. Any disclosure, copying or other use of the information by 
any person or organisation who is not the intended recipient is strictly 
prohibited and may be unlawful. If you have received this e-mail in error, 
please inform the sender immediately and delete/destroy this e-mail and any 
copies of it. Novoda Ltd has taken reasonable precautions to minimise the 
risk of any software viruses which may damage your systems, but we advise 
that you take the necessary steps to ensure that no virus contamination is 
suffered. Novoda Ltd does not accept any liability for any loss or damage 
caused by the transmission of any virus.

Novoda Ltd, Company No: 347444, Registered in Scotland Registered 
Office: C/O Alexander Sloan, 38 Cadogan Street, Glasgow, G2 7HF, Scotland.  VAT 
Registration Number  GB 984 2525 93

File: library/src/main/java/com/novoda/downloadmanager/lib/DownloadStatus.java [hunk 4]
```diff
@@ -225,6 +225,10 @@ public static boolean isDeleting(int status) {
         return status == DELETING;
     }
 
+    public static boolean isPausedByApp(int status) {
```
Comment (ouchadam) [inline library/src/main/java/com/novoda/downloadmanager/lib/DownloadStatus.java]: :+1:

File: library/src/main/java/com/novoda/downloadmanager/lib/DownloadManager.java [hunk 5]
```diff
@@ -445,25 +445,37 @@ private long insert(Request request) {
...
+        RealSystemFacade systemFacade = new RealSystemFacade(GlobalState.getContext(), new Clock());
+        BatchRepository batchRepository = new BatchRepository(contentResolver, downloadDeleter, downloadsUriProvider, systemFacade);
 
-    public void resumeBatch(long id) {
-        ContentValues values = getResumeContentValues();
-        String where = COLUMN_BATCH_ID + "= ? AND " + DownloadContract.Downloads.COLUMN_STATUS + " != ?";
-        String[] selectionArgs = {String.valueOf(id), String.valueOf(DownloadStatus.SUCCESS)};
-        contentResolver.update(downloadsUriProvider.getAllDownloadsUri(), values, where, selectionArgs);
+        int batchStatus = batchRepository.getBatchStatus(batchId);
+        if (DownloadStatus.isRunning(batchStatus)) {
```
Comment (ouchadam) [inline library/src/main/java/com/novoda/downloadmanager/lib/DownloadManager.java]: I think it would be nice if this and the pause method were extracted into a class, but not a blocker for me :+1:

**Comment by ouchadam** [comment]
:+1: LGTM

File: library/src/main/java/com/novoda/downloadmanager/lib/DownloadManager.java (hunk 3 above)
Comment (xrigau) [inline library/src/main/java/com/novoda/downloadmanager/lib/DownloadManager.java]: I was wondering if it'd be ok to throw an exception or something

File: library/src/main/java/com/novoda/downloadmanager/lib/DownloadManager.java (hunk 2 above)
Comment (xrigau) [inline library/src/main/java/com/novoda/downloadmanager/lib/DownloadManager.java]: yep but not part of this PR tho

File: library/src/main/java/com/novoda/downloadmanager/lib/DownloadManager.java (hunk 3 above)
Comment (zegnus) [inline library/src/main/java/com/novoda/downloadmanager/lib/DownloadManager.java]: https://github.com/novoda/download-manager/commit/18083c208a5d0e1a1299d03a2084de3f5bcc4eac

File: library/src/main/java/com/novoda/downloadmanager/lib/DownloadManager.java (hunk 5 above)
Comment (zegnus) [inline library/src/main/java/com/novoda/downloadmanager/lib/DownloadManager.java]: https://github.com/novoda/download-manager/commit/d14e2dce5b16897d262de7dd5d19faad5693960c

**Comment by hal9002** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://ci.novoda.com/job/download-manager-prb/196/
Test PASSed.

File: library/src/main/java/com/novoda/downloadmanager/lib/DownloadManager.java (hunk 3 above)
Comment (amlcurran) [inline library/src/main/java/com/novoda/downloadmanager/lib/DownloadManager.java]: ooh not sure I like the exception - we're not doing anything wrong to try to pause a queued download so throwing an exception seems heavy-handed

**Comment by hal9002** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://ci.novoda.com/job/download-manager-prb/197/
Test PASSed.

**Comment by xrigau** [comment]
unmergeable

**Comment by hal9002** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://ci.novoda.com/job/download-manager-prb/198/
Test PASSed.

**Comment by amlcurran** [comment]
LGTM
